labs/b9.md

```diff
@@ -0,0 +1,178 @@
+ 4. The Docker daemon streamed that output to the Docker client, which sent it to your terminal.    
+
+Some [quick definitions from Docker’s website](https://docs.docker.com/get-started/#a-brief-explanation-of-containers):    
+
+An `image` is a lightweight, stand-alone, executable package that includes everything needed to run a piece of software, including the code, a runtime, libraries, environment variables, and config files.        
+
+A `container` is a runtime instance of an image—what the image becomes in memory when actually executed. It runs completely isolated from the host environment by default, only accessing host files and ports if configured to do so.        
+
+Be sure to read through the output from running the hello-world image to get an understanding of what the Docker daemon was doing.    
+
+#### Running an interactive container    
+
+We’re now going to walk you through running a container interactively. This is useful if you ever need to play around and install stuff on a bare system without messing up your current system. Try running the following command:    
+
+`docker run -it ubuntu:latest`    
+
+The -i flag tells docker to keep `STDIN` open to your container, and the -t flag allocates a [pseudo-TTY](https://en.wikipedia.org/wiki/Pseudoterminal) for you. Basically you need both for you to have a shell into your newly started container. Try installing some packages from apt or just play around. It should look like a bare Linux system.    
+
+You can exit the container with CTRL+D.    
+
+##### Questions    
+
+1.  What user are you logged in as by default?    
+2.  If you start and then exit an interactive container, and then use the `docker run -it ubuntu:latest` command again; is it the same container? How can you tell?      
+
+#### Dockerfiles    
+
+A more powerful way to interface with Docker is by using a `Dockerfile`. A Dockerfile allows you to define an image by specifying all of the commands you would type manually to create an image. Docker can then build images from a specified Dockerfile. These Dockerfiles can be put into version control and the images distributed as a binary to keep track of both how the image is constructed and also to keep pre-built images around.        
+
+Dockerfiles are very powerful and have many different commands and features. We’ll go over a basic example, but you should check out the [reference page](https://docs.docker.com/engine/reference/builder/) if you are trying to do anything more complex.    
+
+Here is an example `Dockerfile` that will build an image based on Fedora Linux that has `python3.6` installed by default. Before building the `Dockerfile`, make sure to first run `docker pull fedora` to retrieve the image from Docker Hub. The code for this exercise can be found in `~/decal-labs/b9/missile`.    
+
+```docker  
+FROM fedora:latest  
+
+RUN yum update -y && yum install python3-pip && \     
```

Comment:
Maybe briefly mention that `yum` is fedora's package manager and the equivalent of `apt`? Maybe also explain a bit about pip, but I feel like that is less relevant and they have probably seen it a bit before.